strfry/plugins/pubkey_cache.ts

```diff
@@ -0,0 +1,59 @@
+// In process cache to avoid accepting pubkeys that were already removed from
+// the relay. We store CACHE_MAX_SIZE items in redis so restarts don't lose the
+// cache. This assumes that after CACHE_MAX_SIZE items are processed, the
+// oldest ones are no longer relevant and probably not used anymore in the
+// network to avoid to have an infinite cache.
+class PubkeyCache {
+  private maxSize: number;
+  private queue: string[] = [];
+  private set: Set<string> = new Set();
```

Comment:
Can you explain why we need a `queue`, a `set`, and a redis cache?

Reply:
I can add comments in the code too if needed:

The `Set` is used for quick O(1) in-process lookups when we need to read from the cache. This ensures that there is very little lag because these lookups are extremely fast, much more than Redis itself. The `Queue` allows us to remove the oldest entries when we exceed the maximum cache size to control memory usage. Both the `Set` and `Queue` are in-process data structures because Strfry blocks until it receives a response from a plugin, so write operations need to be as optimized as possible.

While Redis is great, it’s still too slow to be called on each write operation, it’s asynchronous and remote compared to synchronous local queries, which are orders of magnitude faster. Therefore, the main use for Redis in this case is to persist the cache across restarts. By storing the cache in Redis, we ensure that the application retains important state information even after it restarts, without compromising the performance of critical operations.